File: Tree/377-longestEvenPathNodesCount.cpp
#include "../treeNode.h"
#include<iostream>

int dfs(TreeNode* root,int &val)
{
    if(!root)
        return 0;
    
    int l=dfs(root->left, val);
    int r=dfs(root->right,val);
    if(root->m_val&1==0)  //even node
    {
        val=std::max(val,l+r+1);
        return std::max(l,r)+1;
    }
    return 0;
}

int getLongestEvenPathNodesCount(TreeNode* root)
{
    if(!root)
        return 0;
    
    int ans=0;
    dfs(root,ans);
    return ans;
}

int main()
{
TreeNode *root=new TreeNode(0);
TreeNode *left1=new TreeNode(8);
TreeNode *right1=new TreeNode(2);

TreeNode *rLeft2=new TreeNode(6);
TreeNode *rright2=new TreeNode(0);
TreeNode *rLeft3=new TreeNode(4);

root->left=left1;
root->right=right1;

right1->left=rLeft2;
right1->right=rright2;

rLeft2->left=rLeft3;

std::cout<<getLongestEvenPathNodesCount(root);
}

File: Tree/377-longestEvenPathNodesCount.py
"""
Given a binary tree root, return the longest path consisting of even values between any two nodes in the tree.
"""


class Tree:
    def __init__(self,val,left=None,right=None) -> None:
        self.val=val
        self.left=left
        self.right=right
        
    

class Solution:
    def longestEvenPathNodesCount(self,root):
        if not root:
            return 0
        
        ans=0
        tmpList=[]
        def dfs(root):
            if not root:
                return 0
            nonlocal ans
            l=dfs(root.left)
            r=dfs(root.right)
            if  root.val&1==0:
                ans=max(ans,l+r+1)
                tmpList.append(root.val)
                return 1+max(l,r)
            return 0
            
        dfs(root)
        print(tmpList)
        return ans
    
    
root=Tree(0)
left1=Tree(8)
right1=Tree(2)

rLeft2=Tree(6)
rright2=Tree(0)
rLeft3=Tree(4)

root.left=left1
root.right=right1

right1.left=rLeft2
right1.right=rright2

rLeft2.left=rLeft3

s=Solution()
print(s.longestEvenPathNodesCount(root))
                